example/particles/particles.cpp

```diff
@@ -316,39 +318,173 @@ TaskStatus Defrag(MeshBlock *pmb) {
+
+    particles_update_done = true;
+    for (auto &block : blocks) {
+      // TODO(BRR) Despite this my particles-specific call, this function feels like it
+      // should be generalized
+      auto swarm = block->swarm_data.Get()->Get("my particles");
+      if (!swarm->finished_transport) {
+        particles_update_done = false;
+      }
+    }
+  }
+
+  // Use a more traditional task list for predictable post-MPI evaluations.
+  status = MakeFinalizationTaskCollection().Execute();
+
+  return status;
+}
+
+// TODO(BRR) This should really be in parthenon/src... but it can't just live in Swarm
+// because of the loop over blocks
+TaskStatus StopCommunicationMesh(BlockList_t &blocks) {
+  int num_sent_local = 0;
+  for (auto &block : blocks) {
+    auto &pmb = block;
+    auto sc = pmb->swarm_data.Get();
+    auto swarm = sc->Get("my particles");
+    swarm->finished_transport = false;
+    num_sent_local += swarm->num_particles_sent_;
+  }
+
+  // Boundary transfers on same MPI proc are blocking
+  for (auto &block : blocks) {
+    auto swarm = block->swarm_data.Get()->Get("my particles");
+    for (int n = 0; n < block->pbval->nneighbor; n++) {
+      NeighborBlock &nb = block->pbval->neighbor[n];
+      // TODO(BRR) May want logic like this if we have non-blocking TaskRegions
+      // if (nb.snb.rank != Globals::my_rank) {
+      //  if (swarm->vbswarm->bd_var_.flag[nb.bufid] != BoundaryStatus::completed) {
+      //    printf("[%i] Neighbor %i not complete!\n", Globals::my_rank, n);
+      //    //return TaskStatus::incomplete;
+      //  }
+      //}
+
+      // TODO(BRR) May want to move this logic into a per-cycle initialization call
+      if (swarm->vbswarm->bd_var_.flag[nb.bufid] == BoundaryStatus::completed) {
+        swarm->vbswarm->bd_var_.req_send[nb.bufid] = MPI_REQUEST_NULL;
+      }
+    }
+  }
+
+  int num_sent_global = 0;
+  MPI_Allreduce(&num_sent_local, &num_sent_global, 1, MPI_INT, MPI_SUM, MPI_COMM_WORLD);
+
+  if (num_sent_global == 0) {
+    for (auto &block : blocks) {
+      auto &pmb = block;
+      auto sc = pmb->swarm_data.Get();
+      auto swarm = sc->Get("my particles");
+      swarm->finished_transport = true;
+
+      // TODO(BRR) should this really be done at an initialization step for each cycle?
+      for (int n = 0; n < swarm->vbswarm->bd_var_.nbmax; n++) {
+        auto &nb = pmb->pbval->neighbor[n];
+        swarm->vbswarm->bd_var_.flag[nb.bufid] = BoundaryStatus::waiting;
+      }
+    }
+  }
+
+  for (auto &block : blocks) {
+    auto &pmb = block;
+    auto sc = pmb->swarm_data.Get();
+    auto swarm = sc->Get("my particles");
+    swarm->finished_transport = true;
+    for (int n = 0; n < swarm->vbswarm->bd_var_.nbmax; n++) {
+      auto &nb = pmb->pbval->neighbor[n];
+      swarm->vbswarm->bd_var_.flag[nb.bufid] = BoundaryStatus::waiting;
+    }
+  }
+
+  return TaskStatus::complete;
+}
+
+TaskCollection ParticleDriver::MakeParticlesCreationTaskCollection() {
```

Comment:
Can this method be made const? It's a bit hard for me to follow some of the logic, but from what I can tell Adding the task should not change any of the underlying data structures in the ParticleDriver all that is changed is the task collection which is passed out.

Reply:
Yes I don't think methods like this in general need to alter the state of `ParticleDriver`. Also, it compiles as a `() const {` method.